app/src/main/java/com/codesoom/assignment/network/HttpRouter.java

```diff
@@ -0,0 +1,85 @@
+package com.codesoom.assignment.network;
+
+import com.sun.net.httpserver.HttpExchange;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * 등록된 path, method에 맞는 executor를 실행해줍니다
+ */
+public class HttpRouter {
+
+    private final HashMap<HttpRouterKey, HttpRouterExecutable> pathMap = new HashMap<>();
+
+    /**
+     * GET method에 대한 executor를 등록합니다.
+     * @param pathRegex 매칭에 활용할 path 정규식
+     * @param executor 등록된 method와 path가 일치할 때 executor를 호출합니다
+     */
+    public void get(String pathRegex, HttpRouterExecutable executor) {
+        pathMap.put(new HttpRouterKey(HttpMethod.GET, pathRegex), executor);
+    }
+
+    /**
+     * POST method에 대한 executor를 등록합니다.
+     * @param pathRegex 매칭에 활용할 path 정규식
+     * @param executor 등록된 method와 path가 일치할 때 executor를 호출합니다
+     */
+    public void post(String pathRegex, HttpRouterExecutable executor) {
+        pathMap.put(new HttpRouterKey(HttpMethod.POST, pathRegex), executor);
+    }
+
+    /**
+     * PUT method에 대한 executor를 등록합니다.
+     * @param pathRegex 매칭에 활용할 path 정규식
+     * @param executor 등록된 method와 path가 일치할 때 executor를 호출합니다
+     */
+    public void put(String pathRegex, HttpRouterExecutable executor) {
+        pathMap.put(new HttpRouterKey(HttpMethod.PUT, pathRegex), executor);
+    }
+
+    /**
+     * PATCH method에 대한 executor를 등록합니다.
+     * @param pathRegex 매칭에 활용할 path 정규식
+     * @param executor 등록된 method와 path가 일치할 때 executor를 호출합니다
+     */
+    public void patch(String pathRegex, HttpRouterExecutable executor) {
+        pathMap.put(new HttpRouterKey(HttpMethod.PATCH, pathRegex), executor);
+    }
+
+    /**
+     * DELETE method에 대한 executor를 등록합니다.
+     * @param pathRegex 매칭에 활용할 path 정규식
+     * @param executor 등록된 method와 path가 일치할 때 executor를 호출합니다
+     */
+    public void delete(String pathRegex, HttpRouterExecutable executor) {
+        pathMap.put(new HttpRouterKey(HttpMethod.DELETE, pathRegex), executor);
+    }
+
+    /**
+     * http request를 분석해서 등록된 규칙과 매칭되는 핸들러를 실행합니다.
+     * @param exchange http 요청이 들어왔을 때 전달되는 HttpExchange 객체
+     * @throws IOException responseCode, content를 작성할 때 에러가 발생할 수 있습니다
+     */
+    public void route(HttpExchange exchange) throws IOException {
+        final HttpRequest request = new HttpRequest(exchange);
+        final HttpResponse response = new HttpResponse(exchange);
+
+        if (request.isNotValid()) {
+            response.send(HttpResponseCode.NotFound, null);
+            return;
+        }
+
+        for (Map.Entry<HttpRouterKey, HttpRouterExecutable> entry: pathMap.entrySet()) {
+            final boolean isMatches = entry.getKey().matches(request);
+            if (isMatches) {
```

Comment:
```suggestion
            if (entry.isMatches(request)) {
```

이런 형태를 하면 어떨까요?

이것에 대한 반대의견이 있거나 동의의견이 있으실 것입니다.
어느쪽이던 상관없이 그에대한 자신의 의견을 피력해주세요!

Reply:
여기서 `Map.Entry<HttpRouterKey, HttpRouterExecutable> entry` 라서
`entry.isMatches`을 사용할 수 없습니다ㅜ 

아마도 이런 형태를 의도하셨을 거라 추측하고 수정해봤습니다. [fc3e111](https://github.com/CodeSoom/spring-week1-assignment-1/pull/108/commits/fc3e1115984205e6c2e0d127c7434eb65a97678d)

여기서 고민이 된 부분은 
for loop 안에서 가독성을 위해서 한번 할당되고 지워질 변수를 할당하는게 좋을지
아니면 불필요한 변수 할당을 하지 않고 chaining 해서 사용할지 고민을 했습니다.

수정한 커밋에서는 key, value를 로컬 변수로 할당해봤는데 
비교해보면 기존 보다 가독성이 좋은 것 같아서 푸시했어요!
혹시 다른 의도로 코멘트 해주신거라면 알려주세요

Reply:
안녕하세요! 맞아요. matches 라는 이름이 중복으로 보일 필요가 있을까에 대해서 말씀드린 것이었어요!

Reply:
> 여기서 `Map.Entry<HttpRouterKey, HttpRouterExecutable> entry` 라서 `entry.isMatches`을 사용할 수 없습니다ㅜ
> 
> 아마도 이런 형태를 의도하셨을 거라 추측하고 수정해봤습니다. [fc3e111](https://github.com/CodeSoom/spring-week1-assignment-1/pull/108/commits/fc3e1115984205e6c2e0d127c7434eb65a97678d)
> 
> 여기서 고민이 된 부분은 for loop 안에서 가독성을 위해서 한번 할당되고 지워질 변수를 할당하는게 좋을지 아니면 불필요한 변수 할당을 하지 않고 chaining 해서 사용할지 고민을 했습니다.
> 
> 수정한 커밋에서는 key, value를 로컬 변수로 할당해봤는데 비교해보면 기존 보다 가독성이 좋은 것 같아서 푸시했어요! 혹시 다른 의도로 코멘트 해주신거라면 알려주세요

아주 좋은 질문이에요! 저는 이정도도 괜찮다고 생각해요!
하지만  for-loop에 `HttpRouterKey`, `HttpRouterExecutable` 타입이 있는데 `key`, `value` 라는 변수가 가독성에 의미가 있는가는 고민해보아야할 문제입니다!

또한 
- for-loop을 이용한 request에 맞는 Executable을 가져오는 함수와
- Executable이 존재할 때 execute 하는 함수가 합쳐져 있는데요!

간단하지만 이것도 나누면 더 이해가 쉽지 않을까 생각이 들기도 합니다!

예를들어 다음과 같은 형태입니다.

```java
HttpRouterExecutable executor = matchedExecutor(request);  // entry는 this로 존재하기 때문.
executor.execute(request, response);
```

Reply:
아 이 기능의 본질은 매칭되는 executor를 찾아오는 것이었죠...!
제안 주신 형태가 훨씬 이해하기 좋은 것 같습니다.
제가 HashMap 과 Entry loop 기능 활용에 집중하다보니 핵심을 놓친 것 같네요